cmd/gitserver/server/vcs_syncer_jvm_packages.go

```diff
@@ -0,0 +1,269 @@
+	dependencies, err := s.PackageDependencies(remoteURL.Path)
+	if err != nil {
+		return err
+	}
+
+	for _, dependency := range dependencies {
+		sources, err := coursier.FetchSources(ctx, s.Config, dependency)
+		if err != nil {
+			return err
+		}
+		if len(sources) == 0 {
+			return errors.Errorf("no sources.jar for dependency %s", dependency)
+		}
+	}
+	return nil
+}
+
+// CloneCommand returns the command to be executed for cloning from remote.
+func (s JvmPackagesArtifactSyncer) CloneCommand(ctx context.Context, remoteURL *vcs.URL, bareGitDirectory string) (*exec.Cmd, error) {
+	err := os.MkdirAll(bareGitDirectory, 0755)
+	if err != nil {
+		return nil, err
+	}
+
+	cmd := exec.CommandContext(ctx, "git", "--bare", "init")
+	if _, err := runCommandInDirectory(ctx, cmd, bareGitDirectory); err != nil {
+		return nil, err
+	}
+
+	if err := s.Fetch(ctx, remoteURL, GitDir(bareGitDirectory)); err != nil {
+		return nil, err
+	}
+
+	return exec.CommandContext(ctx, "git", "--version"), nil
+}
+
+// Fetch does nothing for Maven packages because they are immutable and cannot be updated after publishing.
+func (s JvmPackagesArtifactSyncer) Fetch(ctx context.Context, remoteURL *vcs.URL, dir GitDir) error {
+	dependencies, err := s.PackageDependencies(remoteURL.Path)
+	if err != nil {
+		return err
+	}
+
+	tags := make(map[string]bool)
+
+	out, err := runCommandInDirectory(ctx, exec.CommandContext(ctx, "git", "tag"), string(dir))
+	if err != nil {
+		return err
+	}
+
+	for _, line := range strings.Split(out, "\n") {
+		if len(line) == 0 {
+			continue
+		}
+		tags[line] = true
+	}
+
+	for i, dependency := range dependencies {
+		if _, ok := tags[dependency.GitTagFromVersion()]; ok {
+			continue
+		}
+		if err := s.gitPushDependencyTag(ctx, string(dir), dependency, i == 0); err != nil {
+			return errors.Wrapf(err, "error pushing dependency %q", dependency)
+		}
+	}
+
+	for tag := range tags {
+		shouldBeRemoved := true
+		for _, dependency := range dependencies {
+			if dependency.GitTagFromVersion() == tag {
+				shouldBeRemoved = false
+				break
+			}
+		}
+
+		if shouldBeRemoved {
+			cmd := exec.CommandContext(ctx, "git", "tag", "-d", tag)
+			if _, err := runCommandInDirectory(ctx, cmd, string(dir)); err != nil {
+				log15.Error("failed to delete git tag", "error", err, "tag", tag)
+				continue
+			}
+		}
+	}
+
+	return nil
+}
+
+// RemoteShowCommand returns the command to be executed for showing remote.
+func (s JvmPackagesArtifactSyncer) RemoteShowCommand(ctx context.Context, remoteURL *vcs.URL) (cmd *exec.Cmd, err error) {
+	return exec.CommandContext(ctx, "git", "remote", "show", "./"), nil
+}
+
+// PackageDependencies returns the list of JVM dependencies that belong to the given URL path.
+// A URL maps to a single JVM package, which may contain multiple versions (one git tag per version).
+func (s JvmPackagesArtifactSyncer) PackageDependencies(repoUrlPath string) (dependencies []reposource.Dependency, err error) {
```

Comment:
nit: making methods private would be more clear for internal use
```suggestion
func (s JvmPackagesArtifactSyncer) packageDependencies(repoUrlPath string) (dependencies []reposource.Dependency, err error) {
```

Reply:
Nice catch, I think I used this method elsewhere at one point but that's no longer the case 👍